klog_test.go

```diff
@@ -1921,3 +1921,65 @@ func TestStopFlushDaemon(t *testing.T) {
 		t.Error("expected flushD to be stopped")
 	}
 }
+
+func TestBackTrace(t *testing.T) {
+	const skippedNumLines = 2
+	const sizeNumLines = 1
+	const numLinesPerTrace = 2
+
+	_, callerFile, _, _ := runtime.Caller(0)
+	s1 := fmt.Sprintf("%v", Backtrace())
+	normalLines := strings.Count(s1, "\n")
+	if strings.Contains(s1, callerFile) == false {
+		t.Errorf("Expected to see the file, having callsite, in trace %v\n", callerFile)
+	}
+
+	// skip the caller file i.e., the current file where backtrace API is called
+	s2 := fmt.Sprintf("%v", Backtrace(BacktraceSkip(skippedNumLines)))
+	skippedLines := strings.Count(s2, "\n")
+	if strings.Contains(s2, callerFile) == true {
+		t.Errorf("Expected file, having callsite, in trace to be skipped %v\n", callerFile)
+	}
+
+	if normalLines != (skippedLines + (numLinesPerTrace * skippedNumLines)) {
+		t.Errorf("Expected to skip %v set of lines corresponding to Backtrace skip got %v\n", skippedNumLines, skippedLines)
+	}
+
+	//return only the topmost frame
+	s3 := fmt.Sprintf("%v", Backtrace(BacktraceSize(sizeNumLines)))
+	sizeLines := strings.Count(s3, "\n")
+	if strings.Contains(s3, callerFile) == true {
+		t.Errorf("Caller file (%v) not expected when size %v is specified \n", callerFile, sizeNumLines)
+	}
+
+	if sizeLines != (numLinesPerTrace * sizeNumLines) {
+		t.Errorf("Incorrect number of trace lines seen. Expected %v , got %v\n",
+			(numLinesPerTrace * sizeNumLines), sizeLines)
+	}
+
+	//anonymous functions
+	c := func() bool {
+		s := fmt.Sprintf("%v", Backtrace())
+		p1 := regexp.MustCompile("(?s)func1\\(\\).*func2\\(\\).*func3\\(\\)")
+		return p1.MatchString(s)
+	}
+	b := func() bool { return c() }
+	a := func() bool { return b() }
+	if a() == false {
+		t.Error("Expected to see 3 anonymous functions (func1/func2/func3)")
+	}
+
+}
+
+func TestBackTraceAll(t *testing.T) {
+	_, callerFile, _, _ := runtime.Caller(0)
+	s1 := fmt.Sprintf("%v", Backtrace(BacktraceAll(true)))
+	if strings.Contains(s1, callerFile) == false {
+		t.Errorf("Expected to see the file, having callsite, in trace %v\n", callerFile)
+	}
+	// Pattern: goroutine 7 [running]:
+	p := regexp.MustCompile("(?m)(^goroutine [0-9]*.\\[.*\\]*:)")
+	if len(p.FindAllString(s1, -1)) < 2 {
+		t.Error("Expected more than 1 goroutine stack to be printed")
```

Comment:
And what did we get instead?

Test failures should always be written so that they can be debugged without having to run the test under a debugger.